### DISCUSSION TITLE
Is there any plan to add TreeView and ListView to egui?

### DISCUSSION THREAD

**Discussion by ssrlive** [emilk/egui#2999] [General]
I can't find the implements of TreeView or ListView in project. I hope egui contains there widgets. we know it's so impotant in Windows or other OSs.

**Comment by thmxv** [comment]
There is a TreeView demo, in the misc demos: https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/misc_demo_window.rs#L91

For a ListView, I think you just need to use a Layout.

**Reply by hydra** [reply]
this isn't a real tree, it's a set of nested `CollapsingHeaders` which doesn't draw lines to each individual node in the tree.  each nested element only has a vertical line. there are no horizontal lines indicating element. there's also no navigation api for it.

![image](https://github.com/user-attachments/assets/d9fb1193-5d2b-4bf9-a1f5-5b621dc02e88)

**Comment by johnoneil** [comment]
Second the introduction of ListView. This seems like a big omission.

**Comment by oscargus** [comment]
Another benefit of a tree view is that one can then only draw the relevant (visible) information similar to `show_rows` for identically sized rows (which expandable labels are not, and also one will have to consider which are open or not), but taking the tree structure into account.
